C/common/reading_set.cpp

```diff
@@ -192,6 +192,51 @@ ReadingSet::append(const vector<Reading *>& readings)
 	}
 }
 
+/**
+* Deep copy a set of readings to this reading set.
+*/
+bool
+ReadingSet::copy(const ReadingSet& src)
+{
+   vector<Reading *> *readings = new vector<Reading *>;
+   bool copyResult = true;
+   try
+   {
+	   for ( auto rs : src.getAllReadings())
+	   {
+		   std::string assetName = rs->getAssetName();
+		   for ( auto dp : rs->getReadingData())
+		   {
+			   std::string dataPointName  = dp->getName();
+			   DatapointValue dv = dp->getData();
+			   Datapoint *value = new Datapoint(dataPointName, dv);
+			   Reading *in = new Reading(assetName, value);
+			   readings->push_back(in);
+		   }
+	   }
+   }
+   catch (std::bad_alloc& ex)
+   {
+		copyResult = false;
+		readings->clear();
+		Logger::getLogger()->error("Insufficient memory :: failed while copying :%d: reading from ReadingSet :%s: ",readings->size()+1, ex.what());
+   }
+   catch (std::exception& ex)
+   {
+	   copyResult = false;
+	   readings->clear();
+	   Logger::getLogger()->error("Unknown exception :: failed while copying :%d: reading from ReadingSet :%s: ",readings->size()+1, ex.what());
```

Comment:
As above, the +1 is not required and formatting of the message for the user. I am currently putting together guidance/rules for errors, I'll send you a link to the draft.